### PR TITLE
Remove reqwest

### PR DESCRIPTION
**Pull request by mxdvl**
## What does this change?

Remove `reqwest` and favours `fetch`. Fetch is well-supported in browsers, and we already polyfill it through `polyfill.io`. Relies on `node-fetch` for the development build.

This work is mostly done through the conversion of `ajax.js`, to use `fetch`’s API:

| `reqwest(opts)`      | `fetch(url, init?)`    |
|----------------------|------------------------|
| `opts.url`           | `url`                  |
| `opts.method`        | `init.method`          |
| `opts.headers`       | `init.headers` + `new Headers({…})` |
| `opts.data`          | `init.body` + `JSON.stringify(…)`   | 
| `opts.type`          | _not supported_        |
| `opts.contentType`   | `init.headers` + `new Headers({ 'Content-Type': … })`             |
| `opts.crossOrigin`   | `init.mode = 'cors'` + `init.credentials = 'include'` if true     |
| `opts.success`       |  convert to `.then()` and append, but also check for `response.ok`|
| `opts.error`         |  convert to `.catch()` and append   |
| `opts.complete`      |  convert to `.finally()` and append |
| `opts.jsonpCallback` |  _not supported_       |

See [Reqwest’s options][reqwest] and [Fetch’s API][fetch].

[reqwest]: https://github.com/ded/reqwest#options
[fetch]: https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#parameters

## Does this change need to be reproduced in dotcom-rendering ?

- [X] No — **but it will reduce the `commercial.dcr.js` bundle**
- [ ] Yes (please indicate your plans for DCR Implementation)

## What is the value of this and can you measure success?

Smaller commercial bundle, nearly 3kB gzipped.

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [x] No
- [ ] Yes (please give details)

### Does this change break ad-free?

- [x] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Tested

- [x] Locally
- [x] On CODE (optional)

### CODE REVIEW

**Review by gtrufitt — COMMENTED** [review]
Do we need to import the fetch polyfill if we're using polyfill IO with the fetch polyfill requested?

**Comment by mxdvl** [comment]
> @gtrufitt: Do we need to import the fetch polyfill if we're using polyfill IO with the fetch polyfill requested?  

That’s something I wanted to flag as a question in the PR description. Whoops. We would need to have at the very least `node-fetch` as a development dependency for Jest tests to pass.

**Comment by gtrufitt** [comment]
> > @gtrufitt: Do we need to import the fetch polyfill if we're using polyfill IO with the fetch polyfill requested?
> 
> That’s something I wanted to flag as a question in the PR description. Whoops. We would need to have at the very least `node-fetch` as a development dependency for Jest tests to pass.

Yeah, certainly I think we should avoid shipping code to users that they don't require. I think in DCR we use fetch-mock a bit, but I'm not sure how scalable that is to implement in Frontend. node-fetch as a dev dependency seems ok?

File: static/src/javascripts/lib/ajax.js [hunk 1]
```diff
@@ -13,7 +12,7 @@ const ajax = (params) => {
         options.crossOrigin = true;
     }
 
-    const r = reqwest(options);
+    const r = fetch(options);
```
Comment (arelra) [inline static/src/javascripts/lib/ajax.js:15]: I'm wondering if the API of `fetch` is the same as `reqwest` for clients of `ajax.js` ?
For example are all the options passed to `ajax` supported by fetch in here:
https://github.com/guardian/frontend/blob/main/static/src/javascripts/projects/common/modules/identity/api.js

**Review by QuarpT — COMMENTED** [review]
looks good from Identity's side if you're able to confirm @arelra 's question

**Comment by mxdvl** [comment]
It’s a bit of a shame that we still need to support JSON-P. Is this something that’s still needed for the identity API, @QuarpT?

File: static/src/javascripts/projects/common/modules/identity/api.js [hunk 2]
```diff
@@ -279,19 +275,14 @@ export const getSubscribedNewsletters = () => {
...
-        });
-    });
+    fetchJson({
+        url: `${idApiRoot || ''}/users/me/consents`,
+        method: 'PATCH',
+        headers: new Headers({ 'Content-Type': 'application/json'}),
+        credentials: 'include',
+        mode: 'cors',
+        body: JSON.stringify(consents),
+    })
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/identity/api.js:285]: `fetchJson` is now a `Promise`, so we can return it directly.

File: static/src/javascripts/projects/common/modules/identity/upsell/account-creation/AccountCreationForm.js [hunk 3]
```diff
@@ -20,41 +19,43 @@ class AccountCreationForm extends Component {
...
-            },
-            complete: () => {
-                this.setState({ isLoading: false });
-            },
+            } catch (exception) {
+                this.setState({ errors: [genericErrorStr] });
+            }
+        })
+        .finally( () => {
+            this.setState({ isLoading: false });
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/identity/upsell/account-creation/AccountCreationForm.js:58]: This is the biggest change of interface going from `reqwest` to `fetch`.

I think the following mappings are correct, but will require further testing:
- `success` &rarr; `.then()`
- `error: response` &rarr; `.catch(response)`
- `commplete` &rarr; `.finally()`

cc. @guardian/identity

**Review by mxdvl — COMMENTED** [review]
I think this is worth a fresh review.

I’ve remodelled `ajax.js` to use `fetch` rather than request, and replaced the calls in code that makes its way in commercial bundle, which includes the identity api. 👋 @guardian/identity 

I’ve left out the discussions api and avatar api as-is. There’s scope to update these as well as part of this PR, or a subsequent one. 👋 @guardian/discussion

**Comment by mxdvl** [comment]
The identity api was the only place that was using a mix of `ajax` and `fetchJson`, and was using a deprecated interface.

File: static/src/javascripts/lib/ajax.js [hunk 4]
```diff
@@ -13,7 +12,25 @@ const ajax = (params) => {
...
+    const { url } = options;
+    const headers = new Headers();
+    if(options.headers !== undefined) headers.append( options.headers );
+    if(options.contentType !== undefined)
+        headers.append('Content-Type', options.contentType );
+    let headerSize = 0;
+    headers.forEach(() => { headerSize++ });
+
+
+    const initArray = Object.entries({
```
Comment (QuarpT) [inline static/src/javascripts/lib/ajax.js:24]: Is the an advantage for using filter not null with `entries` and `fromEntries` over just setting the keys to undefined?

**Comment by QuarpT** [comment]
looks good! worth testing on CODE with a signed in user
I'll leave the approval to a member of the Dotcom team.

**Comment by mxdvl** [comment]
It has been tested on `CODE` with a fake-ish user (me), and I get correct response codes:

```js
{
  status: "ok"
  user:
  adData: {}
  consents: (16) [{…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}]
  dates: {accountCreatedDate: "2021-03-16T18:03:00Z"}
  hasPassword: true
  id: "200004769"
  primaryEmailAddress: "max.duval+with-ads@gutools.co.uk"
  privateFields: {brazeUuid: "e683f434-497d-44ad-8902-bbbcdcaba1d3", puzzleUuid: "7060e59bf085dd1ef017af026862754cd30e2a382b2b1fbbbd43d5dee6dcacd8", googleTagId: "f329b0c90fc70fb3f374090525c4f773d9c8d2432445aeea8d73ae906f4fc795", legacyPackages: "CRE,RCO", 
  legacyProducts: "CRE,RCO"}
  publicFields: {displayName: "user"}
  statusFields: {userEmailValidated: true}
  userGroups: (2) [{…}, {…}]
}
```

**Review by gtrufitt — APPROVED** [review]
I've had a good read through this and it LGTM. My only comment was whether we could avoid reimplementing a lot between ajax and fetch-json but you're on that in another PR. Nice work!

File: static/src/javascripts/projects/common/modules/identity/api.js [hunk 5]
```diff
@@ -66,14 +64,14 @@ export const getUserFromCookie = () => {
...
-    reqwest({
-        url: `${config.get('page.idApiUrl')}/users/me/newsletters`,
+    fetchJson(`${config.get('page.idApiUrl')}/users/me/newsletters`, {
         method: 'PATCH',
-        type: 'json',
-        contentType: 'application/json',
-        withCredentials: true,
-        crossOrigin: true,
-        data: JSON.stringify(newsletter),
+        headers: new Headers({
```
Comment (oliverlloyd) [inline static/src/javascripts/projects/common/modules/identity/api.js:69]: What does `new Headers` do here? I always thought you could pass a simple object for `headers`

File: static/src/javascripts/projects/common/modules/identity/upsell/account-creation/AccountCreationForm.js [hunk 6]
```diff
@@ -20,41 +19,43 @@ class AccountCreationForm extends Component {
...
+            {
+                method: 'POST',
+                body: JSON.stringify({
+                    csrfToken: this.props.csrfToken,
+                    token: this.props.accountToken,
+                    password: this.state.password,
+                })
+            }
+        )
+        .then( () => {
```
Comment (oliverlloyd) [inline static/src/javascripts/projects/common/modules/identity/upsell/account-creation/AccountCreationForm.js:33]: Not sure how reqwest works but `fetch` will return  a 4** response as success. Correctly so (because the network call was successful) but if the other lib behaves differently this might cause a change in behaviour.

**Review by oliverlloyd — APPROVED** [review]
I had just had one small comment around if we need to check the response from fetch for request errors (4**s). You might have already dealt with this though.

File: static/src/javascripts/projects/common/modules/identity/upsell/account-creation/AccountCreationForm.js (hunk 6 above)
Comment (oliverlloyd) [reply static/src/javascripts/projects/common/modules/identity/upsell/account-creation/AccountCreationForm.js:33]: If it's useful, in DCR we access the `ok` field to [check for errors](https://github.com/guardian/dotcom-rendering/blob/c2dc86aef6b425035d9d669438f2ae5b32660581/src/web/lib/api.tsx#L29)

File: static/src/javascripts/projects/common/modules/identity/api.js (hunk 5 above)
Comment (mxdvl) [reply static/src/javascripts/projects/common/modules/identity/api.js:69]: > **`headers`**
Any headers you want to add to your request, contained within a [`Headers`](https://developer.mozilla.org/en-US/docs/Web/API/Headers) object or an object literal with [`ByteString`](https://developer.mozilla.org/en-US/docs/Web/API/ByteString) values. Note that [some names are forbidden](https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_header_name).

You’re right, I just got carried away with _typing_. I thought it was cleaner/nicer, but happy to keep it as an object literal.

Comment (oliverlloyd) [reply static/src/javascripts/projects/common/modules/identity/api.js]: Oh, there's a Headers object? I learn something new every day around here!

I'm utterly cool with either way, I don't think it matters too much so whatever you think is best is fine with me.

File: static/src/javascripts/projects/common/modules/identity/upsell/account-creation/AccountCreationForm.js (hunk 6 above)
Comment (mxdvl) [reply static/src/javascripts/projects/common/modules/identity/upsell/account-creation/AccountCreationForm.js:33]: That’s a good point. I think this is a good candidate for `fetch-json`, which will throw an error if the status code isn’t `ok`.

File: static/src/javascripts/projects/common/modules/identity/api.js (hunk 5 above)
Comment (mxdvl) [reply static/src/javascripts/projects/common/modules/identity/api.js]: There’s the added complexity of having to import in in node, so reverted to an object literal in 2781ed4

File: static/src/javascripts/lib/ajax.js [hunk 7]
```diff
@@ -13,7 +12,21 @@ const ajax = (params) => {
         options.crossOrigin = true;
     }
 
-    const r = reqwest(options);
+    const { url } = options;
+    const headers = { ...options.headers };
+    if(options.headers !== undefined) headers.append( options.headers );
```
Comment (buck06191) [inline static/src/javascripts/lib/ajax.js:17]: `append` isn't a function on arrays in javascript, only on `ParentNode`. This will probably need changing to 
`if(options.headers !== undefined) headers.push( options.headers );`

Comment (buck06191) [reply static/src/javascripts/lib/ajax.js:17]: Should add that this seemed to stop comments loading for signed in users on live blogs/non-DCR articles

Comment (mxdvl) [reply static/src/javascripts/lib/ajax.js:17]: Absolutely, please see: #23629

**Comment by mxdvl** [comment]
After having a few back-and-forth, we plan to have smaller changes on a feature branch `remove-reqwest-third-time-lucky`.

The first step in this process is #23644.
